### DISCUSSION TITLE
MIDI API selection in Media menu is missing from pdnext

### DISCUSSION THREAD

**Discussion by jamshark70** [sebshader/pdnext#21] [Building/Configuration] [answered]
I found to my surprise this afternoon, in a class, that I couldn't open a MIDI port. (New machine, I had only installed pdnext, not pure-data.)

I could have sworn that the Media menu in pd vanilla has a switch for the MIDI API, but I couldn't find it in pdnext.

So I just downloaded and built pd vanilla (which I don't plan to use anyway, but...). ... And there's "OSS MIDI" and "ALSA MIDI" right there in the Media menu.

AFAICS, then, the only way to configure MIDI in pdnext is to install pd vanilla, configure it there, and then pdnext will pick it up from the .pdsettings file.

EDIT: I spoke too soon... pdnext says "opened 0 MIDI input device(s) and 0 MIDI output device(s)" and there seems to be no way to fix it. That's unfortunate. The color themes were saving my eyes but I sometimes need to use MIDI in class.

**Comment by sebshader** [comment]
@jamshark70 I have only ever been able to compile pd without oss midi support personally. (that includes pd vanilla, not just pd-next).. What flags did you build vanilla with?
you can try compiling with my flags (after installing the alsa & jack header/dev packages, and running ```make clean``` once): 
```./configure --enable-jack CFLAGS=""```
sometimes it seems like these are all necessary:
```./configure --enable-alsa --enable-jack --enable-portaudio --enable-portmidi --disable-oss CFLAGS=""```
for me that doesn't seem to supply any selection of midi api in the media menu though, only audio apis. It just opens alsa midi window if I select "midi preferences". However alsa midi usually works perfectly well.
sometimes the preferences have gotten corrupted before in which case it gets fixed when I start pd with the ```-alsamidi``` flag (but I have never seen any option for OSS MIDI..)

**Comment by sebshader** [comment]
@jamshark70 so if I configure with just ```./configure --enable-alsa --enable-jack CFLAGS=""``` (or without ```--disable-oss``` at least) and then start pd with the ```-ossmidi``` flag once the option shows up from then on.
idk why it doesn't show up by default for pdnext but does with vanilla for you (but not for me..). maybe some difference in how the computer reports midi..

but I think at this point ALSA is far more recommended over OSS anyways..
edit: from what you said it seems like you ONLY had OSS running.. odd. (for me it was just alsa.) in that case you probably just need to start pd with ```-alsamidi``` actually.
Of course, make sure to have the equivalent of the libasound2-dev package installed when configuring/building, and that "ALSA" is in the MIDI API list at the end of configure..

**Comment by jamshark70** [comment]
> edit: from what you said it seems like you ONLY had OSS running

No, I've never used OSS MIDI and I never will. Zero interest in that. I only said that pd vanilla includes an OSS MIDI and ALSA MIDI menu option -- I mentioned OSS first because it appears higher in the menu.

ALSA MIDI. I need interoperability with other ALSA MIDI apps (or connectivity to a2jmidid).

`-alsamidi` does not help.

```
    audio APIs:           PortAudio JACK ALSA OSS 
    midi APIs:            ALSA OSS 
```

libasound2-dev is installed -- should be known because I was able to build pd-vanilla from source and get the correct behavior.

I can only ever get `opened 0 MIDI input device(s) and 0 MIDI output device(s).`.

**Comment by sebshader** [comment]
@jamshark70 so what do you see when you open "Media>MIDI Settings.."?
edit: when I start pd I see "opened alsa MIDI client 128 in:1 out:1" in the console
but I think I may have put in "1" for the number of each of the ports in midi settings

**Comment by jamshark70** [comment]
Pd-vanilla:

![pd-vanilla-media](https://user-images.githubusercontent.com/318301/203072840-3e3030cd-9ad5-4024-a82a-427964c4451b.png)

![pd-vanilla-midi](https://user-images.githubusercontent.com/318301/203072866-f90fc5f1-9147-4f3e-92b0-d64e06b401b1.png)

Pdnext:

![pdnext-media](https://user-images.githubusercontent.com/318301/203072913-3f54acf6-f1ab-4012-ad49-7a53f4fcea3c.png)

![pdnext-midi](https://user-images.githubusercontent.com/318301/203072935-c9267fce-81f9-44fd-8487-c1a1fd232287.png)

I'm quite baffled.

**Comment by sebshader** [comment]
@jamshark70 yes it seems like somehow pdnext isn't getting compiled with alsa.. it isn't in the audio device list either..

because that is the screen I get when I select "OSS" for midi, the top is for alsa (which I do have in pd-next)

**Comment by jamshark70** [comment]
OK then... so how do I switch the configure flags? I do `./configure --enable-jack CFLAGS=""` and nothing gets rebuilt.

Have to make clean or something? Not well documented (the vanilla readme's fault there).

**Comment by sebshader** [comment, marked as answer]
sometimes doing ```autoreconf``` can work before reconfiguring, and ```sudo make clean``` or even ```sudo make distclean``` after reconfiguring but before doing ```make```

**Comment by jamshark70** [comment]
OK that was it, stupid sticky configure options where you changed it but it doesn't mark files for recompilation. (I'm not used to this; SuperCollider uses cmake which is much more aggressive about canceling build products for any configuration change.)

Thanks for your patience!

**Comment by sebshader** [comment]
ok I'll convert this to a discussion because you aren't the 1st person to have issues with needing to ```make clean``` to fix build configuration for midi on linux

**Comment by jamshark70** [comment]
Oh, OK, cool... so where it is right now is: there's a solution -- my last configuration was:

```
./configure --enable-alsa --enable-jack --enable-portaudio --enable-portmidi CFLAGS=""
```

(I had to delete `--disable-oss` because it failed to build that way.)

And the remaining question is: why did pd-vanilla ALSA MIDI "just work" with a simple `./configure --enable-jack` while pdnext requires `--enable-alsa`?

FWIW I'm pretty sure I've seen with the pd-vanilla repository that configuration changes didn't take effect without `make clean` -- while this isn't good behavior IMO, I don't believe it's a pdnext-specific regression.

**Reply by sebshader** [reply]
I have no idea, I had just assumed you had installed the alsa dev headers after configuring pdnext but before configuring vanilla

**Reply by sebshader** [reply]
I would also doubt that it's a regression, wouldn't be surprised if it's been this way for a while.
maybe something to do with recompiling portmidi
